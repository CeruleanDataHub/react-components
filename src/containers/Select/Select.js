import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { Dropdown } from "../Dropdown/Dropdown";

const SelectOption = styled.div`
  padding: 0.8rem 1.25rem;
  cursor: pointer;
  border-bottom: 1px solid #dedede;
  background: ${({ selected }) => (selected ? "#dedede" : "transparent")};

  &:hover {
    background: ${({ selected }) => (selected ? "#d4d4d4" : "#f5f5f5")};
  }
`;

export const Select = ({
  label,
  selectedOption,
  isOpen,
  onOpen,
  onChange,
  items
}) => (
  <Dropdown label={label} isOpen={isOpen} onClick={onOpen}>
    {items.map(item => (
      <SelectOption
        className="select"
        selected={item.id === selectedOption.id}
        onClick={() => onChange(item)}
        key={item.id}
      >
        {item.value}
      </SelectOption>
    ))}
  </Dropdown>
);

Select.propTypes = {
  label: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  selectedOption: PropTypes.object,
  /** Define select visibility state */
  isOpen: PropTypes.bool,
  /** Select options */
  // eslint-disable-next-line react/forbid-prop-types
  items: PropTypes.array,
  /** onOpen handler function */
  onOpen: PropTypes.func,
  /** onChange handler function */
  onChange: PropTypes.func
};

Select.defaultProps = {
  label: "",
  selectedOption: {},
  isOpen: false,
  items: [],
  onOpen: () => {},
  onChange: () => {}
};
