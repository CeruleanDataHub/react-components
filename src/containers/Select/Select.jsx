import PropTypes from "prop-types";
import React, { forwardRef } from "react";
import styled from "styled-components";

const SelectContainer = styled.select`
  font-family: inherit;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid ${({ theme }) => theme.border};
  background: ${({ theme }) => theme.background};
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
`;

const mapOptions = options =>
  options.map(({ id, value, name }) => (
    <option id={id} value={value} key={id}>
      {name || value}
    </option>
  ));

const mapOptionGroups = optionGroups =>
  optionGroups.map(({ group, children }) => (
    <optgroup label={group} key={group}>
      {mapOptions(children)}
    </optgroup>
  ));

export const Select = forwardRef(({ onChange, items, selectedOption }, ref) => {
  const isOptionsGroup = items.some(item =>
    Object.keys(item).includes("group")
  );

  return (
    <SelectContainer onChange={onChange} value={selectedOption} ref={ref}>
      {isOptionsGroup ? mapOptionGroups(items) : mapOptions(items)}
    </SelectContainer>
  );
});

Select.propTypes = {
  /** Currently selected option */
  selectedOption: PropTypes.string,
  /** Select options */
  items: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        value: PropTypes.string,
        name: PropTypes.string
      })
    ),
    PropTypes.arrayOf(
      PropTypes.shape({
        group: PropTypes.string,
        children: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string,
            value: PropTypes.string,
            name: PropTypes.string
          })
        )
      })
    )
  ]),
  /** onChange handler function */
  onChange: PropTypes.func
};

Select.defaultProps = {
  selectedOption: "",
  items: [],
  onChange: () => {}
};
