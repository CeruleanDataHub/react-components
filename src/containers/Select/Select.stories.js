import { withKnobs } from "@storybook/addon-knobs";
import React, { useState } from "react";

import { Typography } from "../Typography/Typography";
import { Select } from "./Select";

export default {
  title: "Select",
  component: Select,
  decorators: [withKnobs]
};

export const SelectStory = () => {
  const items = [
    { id: "1", value: "Label 1" },
    { id: "2", value: "Label 2" },
    {
      id: "3",
      value: "Extra long line of text which will overflow the select container."
    }
  ];

  const [selectedItem, setSelectedItem] = useState(items[0].value);

  return (
    <div style={{ minHeight: 250, width: 300 }}>
      <Typography fontFamily="openSans">
        <Select
          items={items}
          onChange={event => setSelectedItem(event.target.value)}
          selectedOption={selectedItem}
        />
      </Typography>
    </div>
  );
};

export const SelectWithNameStory = () => {
  const items = [
    { id: "1", value: "Label 1", name: "Name 1" },
    { id: "2", value: "Label 2", name: "Name 2" },
    {
      id: "3",
      value:
        "Extra long line of text which will overflow the select container.",
      name:
        "Name 3. Extra long line of text which will overflow the select container."
    }
  ];

  const [selectedItem, setSelectedItem] = useState(items[0].value);

  return (
    <div style={{ minHeight: 250, width: 300 }}>
      <Typography fontFamily="openSans">
        <Select
          items={items}
          onChange={event => setSelectedItem(event.target.value)}
          selectedOption={selectedItem}
        />
      </Typography>
    </div>
  );
};

export const SelectStoryWithOptionGroups = () => {
  const items = [
    {
      group: "Group 1",
      children: [
        { id: "1", value: "Label 1" },
        { id: "2", value: "Label 2" }
      ]
    },
    {
      group: "Group 2",
      children: [
        {
          id: "3",
          value:
            "Extra long line of text which will overflow the select container."
        }
      ]
    }
  ];

  const [selectedItem, setSelectedItem] = useState(items[0].value);

  return (
    <div style={{ minHeight: 250, width: 300 }}>
      <Typography fontFamily="openSans">
        <Select
          items={items}
          onChange={event => setSelectedItem(event.target.value)}
          selectedOption={selectedItem}
        />
      </Typography>
    </div>
  );
};
