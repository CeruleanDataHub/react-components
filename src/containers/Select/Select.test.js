import { mount } from "enzyme";
import React from "react";

import { Select } from "./Select";

describe("Select", () => {
  let component;
  let itemsStub;

  beforeEach(() => {
    itemsStub = [{ id: "1", value: "some value" }];
    component = mount(<Select items={itemsStub} />);
  });

  it("renders", () => {
    expect(component.render()).toMatchSnapshot();
  });

  it("has an option with value as text", () => {
    expect(component.find("option")).toHaveText("some value");
  });

  it("given item with name, has an option with name as text", () => {
    component = mount(
      <Select items={[{ id: "1", value: "some value", name: "some name" }]} />
    );

    expect(component.find("option")).toHaveText("some name");
  });

  it("given selected option, has option selected", () => {
    component = mount(<Select selectedOption="some selected option" />);

    expect(component.find("select")).toHaveValue("some selected option");
  });

  it("given items group, has options grouped", () => {
    const itemsGroupStub = [
      {
        group: "some group",
        children: [
          { id: "1", value: "some value" },
          { id: "2", value: "some other value" }
        ]
      },
      {
        group: "some other group",
        children: [
          {
            id: "3",
            value:
              "Extra long line of text which will overflow the select container."
          }
        ]
      }
    ];

    component = mount(<Select items={itemsGroupStub} />);

    expect(component.find("optgroup")).toExist();
  });

  it("given onChange, when called, calls callback function", () => {
    const onChangeMock = jest.fn();
    component = mount(<Select onChange={onChangeMock} />);

    component.props().onChange();

    expect(onChangeMock).toHaveBeenCalled();
  });

  it("given items with indentation, has an option with correct amount of non-breaking spaces prefixed to the value", () => {
    component = mount(
      <Select items={[{ id: "1", indentLevel: 2, value: "some value" }]} />
    );

    expect(component.find("option")).toHaveText("  some value");
  });
});
