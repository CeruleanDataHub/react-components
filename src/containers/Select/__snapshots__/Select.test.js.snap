// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should be able to call the default onChange function when no onChange callback provided 1`] = `
.c0 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid;
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
}

<ForwardRef
  items={
    Array [
      Object {
        "id": "1",
        "value": "test",
      },
    ]
  }
  onChange={[Function]}
  selectedOption=""
>
  <styled.select
    onChange={[Function]}
    value=""
  >
    <select
      className="c0"
      onChange={[Function]}
      value=""
    >
      <option
        id="1"
        key="1"
        value="test"
      >
        test
      </option>
    </select>
  </styled.select>
</ForwardRef>
`;

exports[`Select should render 1`] = `
.c0 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid;
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
}

<select
  className="c0"
  onChange={[Function]}
  value=""
>
  <option
    id="1"
    value="test"
  >
    test
  </option>
</select>
`;

exports[`Select should render option groups 1`] = `
.c0 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid;
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
}

<select
  className="c0"
  onChange={[Function]}
  value="Label 1"
>
  <optgroup
    label="Group 1"
  >
    <option
      id="1"
      value="Label 1"
    >
      Label 1
    </option>
    <option
      id="2"
      value="Label 2"
    >
      Label 2
    </option>
  </optgroup>
  <optgroup
    label="Group 2"
  >
    <option
      id="3"
      value="Extra long line of text which will overflow the select container."
    >
      Extra long line of text which will overflow the select container.
    </option>
  </optgroup>
</select>
`;

exports[`Select should render selected option 1`] = `
.c0 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid;
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
}

<select
  className="c0"
  onChange={[Function]}
  value="test"
>
  <option
    id="1"
    value="test"
  >
    test
  </option>
</select>
`;

exports[`Select should render with name 1`] = `
.c0 {
  font-family: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 40px;
  border: 1px solid;
  border-radius: 4px;
  padding: 0 3em 0 1.25em;
  cursor: pointer;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  color: #000;
  line-height: 1.5;
}

<select
  className="c0"
  onChange={[Function]}
  value=""
>
  <option
    id="1"
    value="test"
  >
    test with name
  </option>
</select>
`;
